PWGHF/D2H/Tasks/taskDirectedFlowCharmHadrons.cxx

```diff
@@ -0,0 +1,304 @@
+
+    std::vector<AxisSpec> axes = {thnAxisInvMass, thnAxisCent, thnAxisPt, thnAxisEta, thnAxisScalarProd, thnAxisSign};
+    if (storeMl) {
+      axes.insert(axes.end(), {thnAxisMlOne, thnAxisMlTwo});
+    }
+    registry.add("hpuxQxpvscentpteta", "hpuxQxpvscentpteta", HistType::kTHnSparseF, axes, true);
+    registry.add("hpuyQypvscentpteta", "hpuyQypvscentpteta", HistType::kTHnSparseF, axes, true);
+    registry.add("hpuxQxtvscentpteta", "hpuxQxtvscentpteta", HistType::kTHnSparseF, axes, true);
+    registry.add("hpuyQytvscentpteta", "hpuyQytvscentpteta", HistType::kTHnSparseF, axes, true);
+
+    registry.add("hpQxtQxpvscent", "hpQxtQxpvscent", HistType::kTHnSparseF, {thnAxisCent, thnAxisScalarProd}, true);
+    registry.add("hpQytQypvscent", "hpQytQypvscent", HistType::kTHnSparseF, {thnAxisCent, thnAxisScalarProd}, true);
+    registry.add("hpQxtQypvscent", "hpQxtQypvscent", HistType::kTHnSparseF, {thnAxisCent, thnAxisScalarProd}, true);
+    registry.add("hpQxpQytvscent", "hpQxpQytvscent", HistType::kTHnSparseF, {thnAxisCent, thnAxisScalarProd}, true);
+
+    ccdb->setURL(ccdbUrl);
+    ccdb->setCaching(true);
+    ccdb->setLocalObjectValidityChecking();
+  }; // end init
+
+  /// Fill THnSparse
+  /// \param mass is the invariant mass of the candidate
+  /// \param pt is the transverse momentum of the candidate
+  /// \param cent is the centrality of the collision
+  /// \param cosNPhi is the cosine of the n*phi angle
+  /// \param cosDeltaPhi is the cosine of the n*(phi - evtPl) angle
+  /// \param sp is the scalar product
+  /// \param outputMl are the ML scores
+  void fillThn(double& mass,
+               double& pt,
+               double& eta,
+               double& cent,
+               double& uxQxp,
+               double& uyQyp,
+               double& uxQxt,
+               double& uyQyt,
+               std::vector<double>& outputMl,
+               double& sign)
+  {
+    if (storeMl) {
+      registry.fill(HIST("hpuxQxpvscentpteta"), mass, cent, pt, eta, uxQxp, sign, outputMl[0], outputMl[1]);
+      registry.fill(HIST("hpuyQypvscentpteta"), mass, cent, pt, eta, uyQyp, sign, outputMl[0], outputMl[1]);
+      registry.fill(HIST("hpuxQxtvscentpteta"), mass, cent, pt, eta, uxQxt, sign, outputMl[0], outputMl[1]);
+      registry.fill(HIST("hpuyQytvscentpteta"), mass, cent, pt, eta, uyQyt, sign, outputMl[0], outputMl[1]);
+    } else {
+      registry.fill(HIST("hpuxQxpvscentpteta"), mass, cent, pt, eta, uxQxp, sign);
+      registry.fill(HIST("hpuyQypvscentpteta"), mass, cent, pt, eta, uyQyp, sign);
+      registry.fill(HIST("hpuxQxtvscentpteta"), mass, cent, pt, eta, uxQxt, sign);
+      registry.fill(HIST("hpuyQytvscentpteta"), mass, cent, pt, eta, uyQyt, sign);
+    }
+  }
+
+  /// Get the centrality
+  /// \param collision is the collision with the centrality information
+  double getCentrality(CollsWithQvecs::iterator const& collision)
+  {
+    double cent = -999.;
```

Comment:
Can you not reuse the existing utility function with `concepts`?

Reply:
[Reply]: At present would like to keep it as it is.